crates/proof/executor/src/builder/assemble.rs

```diff
@@ -0,0 +1,195 @@
+//! [Header] assembly logic for the [StatelessL2Builder].
+
+use super::StatelessL2Builder;
+use crate::{
+    ExecutorError, ExecutorResult, TrieDBError, TrieDBProvider,
+    constants::{L2_TO_L1_BRIDGE, OUTPUT_ROOT_VERSION, SHA256_EMPTY},
+    util::encode_holocene_eip_1559_params,
+};
+use alloc::vec::Vec;
+use alloy_consensus::{EMPTY_OMMER_ROOT_HASH, Header, Sealed};
+use alloy_eips::Encodable2718;
+use alloy_evm::{EvmFactory, block::BlockExecutionResult};
+use alloy_primitives::{B256, Sealable, U256, hex, keccak256, logs_bloom};
+use alloy_trie::EMPTY_ROOT_HASH;
+use kona_genesis::RollupConfig;
+use kona_mpt::{TrieHinter, ordered_trie_with_encoder};
+use op_alloy_consensus::OpReceiptEnvelope;
+use op_alloy_rpc_types_engine::OpPayloadAttributes;
+use revm::{context::BlockEnv, database::BundleState};
+
+impl<P, H, Evm> StatelessL2Builder<'_, P, H, Evm>
+where
+    P: TrieDBProvider,
+    H: TrieHinter,
+    Evm: EvmFactory,
+{
+    /// Seals the block executed from the given [OpPayloadAttributes] and [BlockEnv], returning the
+    /// computed [Header].
+    pub(crate) fn seal_block(
+        &mut self,
+        attrs: &OpPayloadAttributes,
+        parent_hash: B256,
+        block_env: &BlockEnv,
+        ex_result: &BlockExecutionResult<OpReceiptEnvelope>,
+        bundle: BundleState,
+    ) -> ExecutorResult<Sealed<Header>> {
+        let timestamp = block_env.timestamp;
+
+        // Compute the roots for the block header.
+        let state_root = self.trie_db.state_root(&bundle)?;
+        let transactions_root = ordered_trie_with_encoder(
+            attrs.transactions.as_ref().expect("Transactions must be non-empty"),
```

Comment:
Maybe worthwhile to leave a `// SAFETY: ` comment why it's ok to panic for empty txs

Reply:
Added